File: maze.py
# maze.py
import random
import sys
import os
from collections import deque
import time

class Maze:
    def __init__(self, width=21, height=21):
        self.width = width if width % 2 == 1 else width + 1
        self.height = height if height % 2 == 1 else height + 1
        self.grid = []
        self.start = (1, 1)
        self.end = (self.height - 2, self.width - 2)
        self.generate()
        self.solution = []

    def generate(self):
        # Initialize grid with all walls
        self.grid = [['#'] * self.width for _ in range(self.height)]
        # Carve paths using DFS
        stack = [(1, 1)]
        self.grid[1][1] = ' '
        dirs = [(0, 2), (0, -2), (2, 0), (-2, 0)]
        while stack:
            cx, cy = stack[-1]
            neighbours = []
            for dx, dy in dirs:
                nx, ny = cx + dx, cy + dy
                if 0 < nx < self.height and 0 < ny < self.width and self.grid[nx][ny] == '#':
                    neighbours.append((nx, ny, dx//2, dy//2))
            if neighbours:
                nx, ny, wx, wy = random.choice(neighbours)
                self.grid[cx + wx][cy + wy] = ' '
                self.grid[nx][ny] = ' '
                stack.append((nx, ny))
            else:
                stack.pop()
        # Set start and end
        self.grid[self.start[0]][self.start[1]] = 'S'
        self.grid[self.end[0]][self.end[1]] = 'E'

    def display(self):
        for row in self.grid:
            print(''.join(row))

    def solve(self):
        # BFS from start to end
        start = self.start
        end = self.end
        queue = deque([start])
        visited = {start}
        parent = {start: None}
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        while queue:
            cx, cy = queue.popleft()
            if (cx, cy) == end:
                break
            for dx, dy in dirs:
                nx, ny = cx + dx, cy + dy
                if 0 <= nx < self.height and 0 <= ny < self.width:
                    if self.grid[nx][ny] != '#' and (nx, ny) not in visited:
                        visited.add((nx, ny))
                        parent[(nx, ny)] = (cx, cy)
                        queue.append((nx, ny))
        # Reconstruct path
        if end not in parent:
            return []
        path = []
        cur = end
        while cur:
            path.append(cur)
            cur = parent[cur]
        path.reverse()
        self.solution = path
        return path

    def display_solution(self):
        if not self.solution:
            self.solve()
        display = [list(row) for row in self.grid]
        for (x, y) in self.solution[1:-1]:
            if display[x][y] not in ('S', 'E'):
                display[x][y] = 'o'
        for row in display:
            print(''.join(row))

    def save(self, filename):
        with open(filename, 'w') as f:
            for row in self.grid:
                f.write(''.join(row) + '\n')

    def load(self, filename):
        with open(filename, 'r') as f:
            self.grid = [list(line.strip()) for line in f if line.strip()]
        self.height = len(self.grid)
        self.width = len(self.grid[0])
        for i in range(self.height):
            for j in range(self.width):
                if self.grid[i][j] == 'S':
                    self.start = (i, j)
                elif self.grid[i][j] == 'E':
                    self.end = (i, j)
        self.solution = []

def main():
    w = int(sys.argv[1]) if len(sys.argv) > 1 else 21
    h = int(sys.argv[2]) if len(sys.argv) > 2 else 21
    maze = Maze(w, h)
    print("🧩 Maze Generator")
    print("Commands: generate, solve, save <file>, load <file>, quit")
    while True:
        cmd = input("> ").strip().split()
        if not cmd:
            continue
        if cmd[0] == 'quit':
            break
        elif cmd[0] == 'generate':
            maze = Maze(w, h)
            maze.display()
        elif cmd[0] == 'solve':
            maze.display_solution()
        elif cmd[0] == 'save' and len(cmd) > 1:
            maze.save(cmd[1])
            print(f"Saved to {cmd[1]}")
        elif cmd[0] == 'load' and len(cmd) > 1:
            maze.load(cmd[1])
            print(f"Loaded from {cmd[1]}")
            maze.display()
        else:
            print("Unknown command.")

if __name__ == "__main__":
    main()

File: maze.rb
# maze.rb
require 'set'

class Maze
  attr_reader :grid, :width, :height, :start, :end, :solution

  def initialize(w = 21, h = 21)
    @width = w.even? ? w + 1 : w
    @height = h.even? ? h + 1 : h
    @start = [1, 1]
    @end = [@height - 2, @width - 2]
    @solution = nil
    generate
  end

  def generate
    @grid = Array.new(@height) { Array.new(@width, '#') }
    @grid[1][1] = ' '
    stack = [[1, 1]]
    dirs = [[0, 2], [0, -2], [2, 0], [-2, 0]]
    until stack.empty?
      cx, cy = stack.last
      neighbours = []
      dirs.each do |dx, dy|
        nx, ny = cx + dx, cy + dy
        if nx > 0 && nx < @height - 1 && ny > 0 && ny < @width - 1 && @grid[nx][ny] == '#'
          neighbours << [nx, ny, dx / 2, dy / 2]
        end
      end
      if neighbours.any?
        nx, ny, wx, wy = neighbours.sample
        @grid[cx + wx][cy + wy] = ' '
        @grid[nx][ny] = ' '
        stack << [nx, ny]
      else
        stack.pop
      end
    end
    @grid[@start[0]][@start[1]] = 'S'
    @grid[@end[0]][@end[1]] = 'E'
    @solution = nil
  end

  def display
    @grid.each { |row| puts row.join }
  end

  def solve
    queue = [@start]
    visited = Set.new([@start])
    parent = {}
    dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]]
    until queue.empty?
      cx, cy = queue.shift
      break if [cx, cy] == @end
      dirs.each do |dx, dy|
        nx, ny = cx + dx, cy + dy
        if nx >= 0 && nx < @height && ny >= 0 && ny < @width && @grid[nx][ny] != '#'
          unless visited.include?([nx, ny])
            visited << [nx, ny]
            parent[[nx, ny]] = [cx, cy]
            queue << [nx, ny]
          end
        end
      end
    end
    if parent[[@end[0], @end[1]]].nil? && @end != @start
      @solution = nil
      return nil
    end
    path = []
    cur = @end
    while cur
      path.unshift(cur)
      break if cur == @start
      cur = parent[cur]
    end
    @solution = path
    path
  end

  def display_solution
    solve if @solution.nil?
    display_grid = @grid.map(&:dup)
    @solution[1...-1].each do |x, y|
      display_grid[x][y] = 'o' unless display_grid[x][y] == 'S' || display_grid[x][y] == 'E'
    end
    display_grid.each { |row| puts row.join }
  end

  def save(filename)
    File.open(filename, 'w') do |f|
      @grid.each { |row| f.puts row.join }
    end
  end

  def load(filename)
    lines = File.readlines(filename).map(&:chomp).reject(&:empty?)
    @grid = lines.map(&:chars)
    @height = @grid.size
    @width = @grid[0].size
    @height.times do |i|
      @width.times do |j|
        @start = [i, j] if @grid[i][j] == 'S'
        @end = [i, j] if @grid[i][j] == 'E'
      end
    end
    @solution = nil
  end
end

def main
  w = (ARGV[0] || 21).to_i
  h = (ARGV[1] || 21).to_i
  maze = Maze.new(w, h)
  puts "🧩 Maze Generator"
  puts "Commands: generate, solve, save <file>, load <file>, quit"
  loop do
    print "> "
    parts = gets.chomp.strip.split
    break if parts.empty?
    case parts[0]
    when 'quit'
      break
    when 'generate'
      maze = Maze.new(w, h)
      maze.display
    when 'solve'
      maze.display_solution
    when 'save'
      if parts.size > 1
        maze.save(parts[1])
        puts "Saved to #{parts[1]}"
      end
    when 'load'
      if parts.size > 1
        begin
          maze.load(parts[1])
          puts "Loaded from #{parts[1]}"
          maze.display
        rescue => e
          puts "Error loading file."
        end
      end
    else
      puts "Unknown command."
    end
  end
end

main if __FILE__ == $0
